### PR TITLE
Refactor a part of PerfEventReaders

### PR DESCRIPTION
**Pull request by Mahmood-Darwish**
This PR refactors some of the consumer functions in PerfEventReaders to use the new ConsumeRecordSample.

Part of bug: http://b/241339791

Tests: Manual tests (compared scheduler view, tracepoint info, and unwinding error percentages before and after the change on a gamelet) + Ran LinuxTracingIntegrationTests

### CODE REVIEW

**Comment by Mahmood-Darwish** [comment]
@beckerhe  Should there be a deconstructor for PerfRecordSample at the end of the consumer functions?

File: src/LinuxTracing/PerfEventReaders.cpp [hunk 1]
```diff
@@ -295,128 +296,55 @@ MmapPerfEvent ConsumeMmapPerfEvent(PerfEventRingBuffer* ring_buffer,
 
 StackSamplePerfEvent ConsumeStackSamplePerfEvent(PerfEventRingBuffer* ring_buffer,
                                                  const perf_event_header& header) {
-  // We expect the following layout of the perf event:
-  //  struct {
-  //    struct perf_event_header header;
```
Comment (karupayun) [inline src/LinuxTracing/PerfEventReaders.cpp:300]: Are you sure about erasing this comment? Why not just erasing the last part of the computation of offsets?

File: src/LinuxTracing/PerfEventReaders.cpp [hunk 2]
```diff
@@ -295,128 +296,55 @@ MmapPerfEvent ConsumeMmapPerfEvent(PerfEventRingBuffer* ring_buffer,
...
-
-  size_t offset_of_dyn_size = offset_of_data + (size * sizeof(char));
-
-  uint64_t dyn_size = 0;
-  ring_buffer->ReadValueAtOffset(&dyn_size, offset_of_dyn_size);
+  perf_event_attr flags{};
 
-  perf_event_sample_id_tid_time_streamid_cpu sample_id;
-  ring_buffer->ReadValueAtOffset(&sample_id, offsetof(perf_event_stack_sample_fixed, sample_id));
+  flags.sample_type =
```
Comment (karupayun) [inline src/LinuxTracing/PerfEventReaders.cpp:301]: Can I ask for a comment here about why these 3 are used for sample_type?

File: src/LinuxTracing/PerfEventReaders.cpp [hunk 3]
```diff
@@ -295,128 +296,55 @@ MmapPerfEvent ConsumeMmapPerfEvent(PerfEventRingBuffer* ring_buffer,
...
-                                 offsetof(perf_event_stack_sample_fixed, regs));
-  ring_buffer->ReadRawAtOffset(event.data.data.get(), offset_of_data, dyn_size);
   ring_buffer->SkipRecord(header);
   return event;
 }
 
 CallchainSamplePerfEvent ConsumeCallchainSamplePerfEvent(PerfEventRingBuffer* ring_buffer,
                                                          const perf_event_header& header) {
-  // We expect the following layout of the perf event:
-  //  struct {
```
Comment (karupayun) [inline src/LinuxTracing/PerfEventReaders.cpp:358]: Similar to above:

Are you sure about erasing this comment? Why not just erasing the last part of the computation of offsets?

File: src/LinuxTracing/PerfEventReaders.cpp [hunk 4]
```diff
@@ -295,128 +296,55 @@ MmapPerfEvent ConsumeMmapPerfEvent(PerfEventRingBuffer* ring_buffer,
...
-  //    u64 dyn_size;           /* if PERF_SAMPLE_STACK_USER && size != 0 */
-  //  };
-  // Unfortunately, the number of `ips` is dynamic, so we need to compute the offsets by hand,
-  // rather than relying on a struct.
-  uint64_t nr = 0;
-  ring_buffer->ReadValueAtOffset(&nr, offsetof(perf_event_callchain_sample_fixed, nr));
+  perf_event_attr flags{};
 
-  const uint64_t size_of_ips_in_bytes = nr * sizeof(uint64_t);
+  flags.sample_type = PERF_SAMPLE_REGS_USER | PERF_SAMPLE_STACK_USER | PERF_SAMPLE_CALLCHAIN |
```
Comment (karupayun) [inline src/LinuxTracing/PerfEventReaders.cpp:328]: Similar to above, Can I ask for a comment here about why these 4 are used for sample_type?

File: src/LinuxTracing/PerfEventReaders.cpp [hunk 5]
```diff
@@ -479,16 +407,21 @@ UprobesWithStackPerfEvent ConsumeUprobeWithStackPerfEvent(PerfEventRingBuffer* r
 
 GenericTracepointPerfEvent ConsumeGenericTracepointPerfEvent(PerfEventRingBuffer* ring_buffer,
                                                              const perf_event_header& header) {
-  perf_event_raw_sample_fixed ring_buffer_record;
-  ring_buffer->ReadRawAtOffset(&ring_buffer_record, 0, sizeof(perf_event_raw_sample_fixed));
+  perf_event_attr flags{};
+
+  flags.sample_type = SAMPLE_TYPE_TID_TIME_STREAMID_CPU;
```
Comment (karupayun) [inline src/LinuxTracing/PerfEventReaders.cpp:412]: Similar to above, Can I ask for a comment here about why SAMPLE_TYPE_TID_TIME_... is the only needed for sample_type?

File: src/LinuxTracing/PerfEventReaders.cpp [hunk 6]
```diff
@@ -498,31 +431,26 @@ GenericTracepointPerfEvent ConsumeGenericTracepointPerfEvent(PerfEventRingBuffer
 
 SchedWakeupPerfEvent ConsumeSchedWakeupPerfEvent(PerfEventRingBuffer* ring_buffer,
                                                  const perf_event_header& header) {
-  ORBIT_CHECK(header.size >= sizeof(perf_event_raw_sample_fixed));
-  perf_event_raw_sample_fixed ring_buffer_record;
-  ring_buffer->ReadRawAtOffset(&ring_buffer_record, 0, sizeof(perf_event_raw_sample_fixed));
+  perf_event_attr flags{};
+
+  flags.sample_type = PERF_SAMPLE_RAW | SAMPLE_TYPE_TID_TIME_STREAMID_CPU;
+
```
Comment (karupayun) [inline src/LinuxTracing/PerfEventReaders.cpp:437]: Aren't you missing:

`  flags.sample_regs_user = SAMPLE_REGS_USER_ALL;`

File: src/LinuxTracing/PerfEventReaders.cpp [hunk 7]
```diff
@@ -498,31 +431,26 @@ GenericTracepointPerfEvent ConsumeGenericTracepointPerfEvent(PerfEventRingBuffer
 
 SchedWakeupPerfEvent ConsumeSchedWakeupPerfEvent(PerfEventRingBuffer* ring_buffer,
                                                  const perf_event_header& header) {
-  ORBIT_CHECK(header.size >= sizeof(perf_event_raw_sample_fixed));
-  perf_event_raw_sample_fixed ring_buffer_record;
-  ring_buffer->ReadRawAtOffset(&ring_buffer_record, 0, sizeof(perf_event_raw_sample_fixed));
+  perf_event_attr flags{};
+
+  flags.sample_type = PERF_SAMPLE_RAW | SAMPLE_TYPE_TID_TIME_STREAMID_CPU;
```
Comment (karupayun) [inline src/LinuxTracing/PerfEventReaders.cpp:436]: Similar to above, Can I ask for a comment here about why these 2 are the only needed for sample_type?

File: src/LinuxTracing/PerfEventReaders.cpp [hunk 8]
```diff
@@ -498,31 +431,26 @@ GenericTracepointPerfEvent ConsumeGenericTracepointPerfEvent(PerfEventRingBuffer
...
-  // The last fields of the sched:sched_wakeup tracepoint aren't always the same, depending on the
-  // kernel version. Fortunately we only need the first fields, which are always the same, so only
-  // read those. See `sched_wakeup_tracepoint_fixed`.
-  ORBIT_CHECK(ring_buffer_record.size >= sizeof(sched_wakeup_tracepoint_fixed));
   sched_wakeup_tracepoint_fixed sched_wakeup;
-  ring_buffer->ReadRawAtOffset(
-      &sched_wakeup,
-      offsetof(perf_event_raw_sample_fixed, size) + sizeof(perf_event_raw_sample_fixed::size),
-      sizeof(sched_wakeup_tracepoint_fixed));
+  memcpy(&sched_wakeup, res.raw_data.get(), sizeof(sched_wakeup_tracepoint_fixed));
```
Comment (karupayun) [inline src/LinuxTracing/PerfEventReaders.cpp:441]: Maybe a stupid question. But why you are only using memcpy here and not in the other methods?

**Review by karupayun — COMMENTED** [review]
Hey Mahmood, I'm not very familiar with PerfEventReaders but I left a few comments! :)

File: src/LinuxTracing/PerfEventReaders.cpp (hunk 8 above)
Comment (beckerhe) [inline src/LinuxTracing/PerfEventReaders.cpp:441]: Nit: `memcpy` -> `std::memcpy`

File: src/LinuxTracing/PerfEventReaders.cpp [hunk 9]
```diff
@@ -479,16 +407,21 @@ UprobesWithStackPerfEvent ConsumeUprobeWithStackPerfEvent(PerfEventRingBuffer* r
 
 GenericTracepointPerfEvent ConsumeGenericTracepointPerfEvent(PerfEventRingBuffer* ring_buffer,
                                                              const perf_event_header& header) {
-  perf_event_raw_sample_fixed ring_buffer_record;
-  ring_buffer->ReadRawAtOffset(&ring_buffer_record, 0, sizeof(perf_event_raw_sample_fixed));
+  perf_event_attr flags{};
+
+  flags.sample_type = SAMPLE_TYPE_TID_TIME_STREAMID_CPU;
+  flags.sample_regs_user = SAMPLE_REGS_USER_ALL;
```
Comment (beckerhe) [inline src/LinuxTracing/PerfEventReaders.cpp:413]: I don't see this being used. Is that intended?

**Review by beckerhe — COMMENTED** [review]
Nice. Can you add to the pull request description how you tested this?

File: src/LinuxTracing/PerfEventReaders.cpp (hunk 1 above)
Comment (Mahmood-Darwish) [reply src/LinuxTracing/PerfEventReaders.cpp:300]: I don't think they are needed anymore. At first the functions were really messy and you needed to see the struct to understand what was going on but now you should be able to understand what's happening just by looking at the flags. The comment doesn't add much info and it clutters the view. Besides we already have PerfEventRecord as an actual struct in the code, if you want you can just look at the flags and then look at the struct and you'll understand perfectly what's being populated.

File: src/LinuxTracing/PerfEventReaders.cpp (hunk 2 above)
Comment (Mahmood-Darwish) [reply src/LinuxTracing/PerfEventReaders.cpp:301]: Those are the fields used by `StackSamplePerfEvent`. Basically, the flags follow the return type. 

`StackSamplePerfEvent` needs time, pid, tid, those are returned by `SAMPLE_TYPE_TID_TIME_STREAMID_CPU`. It also needs regs, that's returned by `PERF_SAMPLE_REGS_USER`. And finally, it needs dyn_size, and data which are returned by `PERF_SAMPLE_STACK_USER`.

To me, they look clear. Should I add a comment about this in the code?

Comment (Mahmood-Darwish) [reply src/LinuxTracing/PerfEventReaders.cpp:301]: Looking at `PerfRecordSample` makes it clear what each flag is responsible for.

File: src/LinuxTracing/PerfEventReaders.cpp (hunk 6 above)
Comment (Mahmood-Darwish) [reply src/LinuxTracing/PerfEventReaders.cpp:437]: Nope, we only need to specify `sample_regs_user` if we set `PERF_SAMPLE_REGS_USER` flag in `sample_type`.

File: src/LinuxTracing/PerfEventReaders.cpp (hunk 8 above)
Comment (Mahmood-Darwish) [reply src/LinuxTracing/PerfEventReaders.cpp:441]: That data we get from `PERF_SAMPLE_RAW` is unstructured and is just an array of bytes.  `sched_wakeup_tracepoint_fixed` gives representation for that data and we copy it there to make it easier to work with. 
We need a `memcpy` everywhere we set `PERF_SAMPLE_RAW`.

File: src/LinuxTracing/PerfEventReaders.cpp [hunk 10]
```diff
@@ -295,128 +296,55 @@ MmapPerfEvent ConsumeMmapPerfEvent(PerfEventRingBuffer* ring_buffer,
...
-
-  uint64_t dyn_size = 0;
-  ring_buffer->ReadValueAtOffset(&dyn_size, offset_of_dyn_size);
+  perf_event_attr flags{};
 
-  perf_event_sample_id_tid_time_streamid_cpu sample_id;
-  ring_buffer->ReadValueAtOffset(&sample_id, offsetof(perf_event_stack_sample_fixed, sample_id));
+  flags.sample_type =
+      PERF_SAMPLE_REGS_USER | PERF_SAMPLE_STACK_USER | SAMPLE_TYPE_TID_TIME_STREAMID_CPU;
+  flags.sample_regs_user = SAMPLE_REGS_USER_ALL;
```
[the remark this answers is not in the document]
Comment (Mahmood-Darwish) [reply src/LinuxTracing/PerfEventReaders.cpp]: Done.

File: src/LinuxTracing/PerfEventReaders.cpp (hunk 8 above)
Comment (Mahmood-Darwish) [reply src/LinuxTracing/PerfEventReaders.cpp]: Done.

File: src/LinuxTracing/PerfEventReaders.cpp (hunk 9 above)
Comment (Mahmood-Darwish) [reply src/LinuxTracing/PerfEventReaders.cpp]: No, it's not. That's a mistake. I removed it now. It should be solved.

File: src/LinuxTracing/PerfEventReaders.cpp (hunk 2 above)
Comment (karupayun) [reply src/LinuxTracing/PerfEventReaders.cpp]: We agreed offline about a comment pointing to PerfEventOpen file and also adding a TODO about not having duplicate code.

**Comment by beckerhe** [comment]
> @beckerhe Should there be a deconstructor for PerfRecordSample at the end of the consumer functions?

Discussed offline. No need for special handling here.

File: src/LinuxTracing/PerfEventReaders.cpp (hunk 2 above)
Comment (Mahmood-Darwish) [reply src/LinuxTracing/PerfEventReaders.cpp]: Added.

File: src/LinuxTracing/PerfEventReaders.cpp [hunk 11]
```diff
@@ -295,128 +296,61 @@ MmapPerfEvent ConsumeMmapPerfEvent(PerfEventRingBuffer* ring_buffer,
...
-  ring_buffer->ReadValueAtOffset(&size, offset_of_size);
-
-  size_t offset_of_dyn_size = offset_of_data + (size * sizeof(char));
-
-  uint64_t dyn_size = 0;
-  ring_buffer->ReadValueAtOffset(&dyn_size, offset_of_dyn_size);
-
-  perf_event_sample_id_tid_time_streamid_cpu sample_id;
-  ring_buffer->ReadValueAtOffset(&sample_id, offsetof(perf_event_stack_sample_fixed, sample_id));
+  // The flags here should be in sync with stack_sample_event_open in
```
Comment (karupayun) [inline src/LinuxTracing/PerfEventReaders.cpp:299]: I don't like "should". They are in sync currently, am I wrong? I would say that the _flags are in sync_.

File: src/LinuxTracing/PerfEventReaders.cpp [hunk 12]
```diff
@@ -295,128 +296,61 @@ MmapPerfEvent ConsumeMmapPerfEvent(PerfEventRingBuffer* ring_buffer,
...
+  // PerfEventOpen.
+  // TODO(): use the same perf_event_attr object from stack_sample_event_open
```
Comment (karupayun) [inline src/LinuxTracing/PerfEventReaders.cpp:301]: Please create a bug and add the link here

**Review by beckerhe — APPROVED** [review]

File: src/LinuxTracing/PerfEventReaders.cpp (hunk 11 above)
Comment (Mahmood-Darwish) [reply src/LinuxTracing/PerfEventReaders.cpp:299]: They are currently in sync, I was thinking more in the sense that if someone wants to change the function they should make sure that the flags are in sync.

I will change the comment.

File: src/LinuxTracing/PerfEventReaders.cpp (hunk 12 above)
Comment (Mahmood-Darwish) [reply src/LinuxTracing/PerfEventReaders.cpp]: Done!

**Review by karupayun — APPROVED** [review]
Thanks for the comments :)

File: src/LinuxTracing/PerfEventReaders.cpp [hunk 13]
```diff
@@ -295,128 +296,59 @@ MmapPerfEvent ConsumeMmapPerfEvent(PerfEventRingBuffer* ring_buffer,
...
-  uint64_t dyn_size = 0;
-  ring_buffer->ReadValueAtOffset(&dyn_size, offset_of_dyn_size);
-
-  perf_event_sample_id_tid_time_streamid_cpu sample_id;
-  ring_buffer->ReadValueAtOffset(&sample_id, offsetof(perf_event_stack_sample_fixed, sample_id));
+  // The flags here are in sync with stack_sample_event_open in PerfEventOpen.
+  // TODO(b/242020362): use the same perf_event_attr object from stack_sample_event_open
+  const perf_event_attr flags{
+      .sample_type =
+          PERF_SAMPLE_REGS_USER | PERF_SAMPLE_STACK_USER | SAMPLE_TYPE_TID_TIME_STREAMID_CPU,
```
Comment (florian-kuebler) [inline src/LinuxTracing/PerfEventReaders.cpp:303]: Nit: we might also want to set the user stack dump size for consistency (also below). But also not needed and will be obsolete once moving to use the actual perf_event_attr.

File: src/LinuxTracing/PerfEventReaders.cpp [hunk 14]
```diff
@@ -479,16 +411,22 @@ UprobesWithStackPerfEvent ConsumeUprobeWithStackPerfEvent(PerfEventRingBuffer* r
 
 GenericTracepointPerfEvent ConsumeGenericTracepointPerfEvent(PerfEventRingBuffer* ring_buffer,
                                                              const perf_event_header& header) {
-  perf_event_raw_sample_fixed ring_buffer_record;
-  ring_buffer->ReadRawAtOffset(&ring_buffer_record, 0, sizeof(perf_event_raw_sample_fixed));
+  // The flags here are in sync with generic_event_attr in PerfEventOpen.
+  // TODO(b/242020362): use the same perf_event_attr object from generic_event_attr
+  const perf_event_attr flags{
+      .sample_type = SAMPLE_TYPE_TID_TIME_STREAMID_CPU,
```
Comment (florian-kuebler) [inline src/LinuxTracing/PerfEventReaders.cpp:417]: Wait, is that correct? Shouldn't we always set the SAMPLE_RAW?

Comment (Mahmood-Darwish) [reply src/LinuxTracing/PerfEventReaders.cpp:417]: Yes, it's correct. This function only reads the fields that are common between all tracepoints. It doesn't read tracepoint specific fields. Reading SAMPLE_RAW will differ based on what type of tracepoint you're trying to read.

File: src/LinuxTracing/PerfEventReaders.cpp (hunk 13 above)
Comment (Mahmood-Darwish) [reply src/LinuxTracing/PerfEventReaders.cpp:303]: Why would we set the stack size here? It won't be used. We will know the stack size we will read from the ring_buffer. One of the fields that we get is the stack size.

File: src/LinuxTracing/PerfEventReaders.cpp (hunk 14 above)
Comment (florian-kuebler) [reply src/LinuxTracing/PerfEventReaders.cpp:417]: Thanks for checking. So I guess the "generic trace points" are used for the trace point track, were we do not show the actual tracepoint data, but only a timestamp.

**Review by florian-kuebler — APPROVED** [review]

**Pull request merged**